src/export/ExportFileDialog.cpp

```diff
@@ -0,0 +1,345 @@
+#include <wx/filectrl.h>
+
+wxDEFINE_EVENT(AUDACITY_FILE_SUFFIX_EVENT, wxCommandEvent);
+
+BEGIN_EVENT_TABLE(ExportFileDialog, wxEvtHandler)
+   EVT_BUTTON(wxID_HELP, ExportFileDialog::OnHelp)
+   EVT_COMMAND(wxID_ANY, AUDACITY_FILE_SUFFIX_EVENT, ExportFileDialog::OnExtensionChanged)
+END_EVENT_TABLE()
+
+namespace {
+
+bool IsExtension(const ExportPlugin& plugin, int formatIndex, const FileExtension& ext)
+{
+   if(formatIndex >= 0 && formatIndex < plugin.GetFormatCount())
+   {
+      auto formatInfo = plugin.GetFormatInfo(formatIndex);
+      if(formatInfo.mExtensions[0].empty() || formatInfo.mExtensions.Index(ext, false) != wxNOT_FOUND)
+         return true;
+   }
+   return false;
+}
+
+}
+
+ExportFileDialog::ExportFileDialog(wxWindow *parent,
+                                   Exporter& exporter,
+                                   const wxString& defaultDir,
+                                   const wxString& defaultName,
+                                   const wxString& defaultFormatName,
+                                   const TranslatableString& title,
+                                   const wxPoint& pos,
+                                   const wxSize& sz,
+                                   const wxString& name)
+   : mExporter(exporter)
+{
+   FileNames::FileTypes fileTypes;
+   
+   const auto& plugins = exporter.GetPlugins();
+   
+   int pluginIndex{0};
+   int formatIndex{0};
+   
+   {
+      int i = 0;
+      for (const auto &plugin : plugins)
+      {
+         for (int j = 0; j < plugin->GetFormatCount(); j++)
+         {
+            auto formatInfo = plugin->GetFormatInfo(j);
+            fileTypes.insert(fileTypes.end(), { formatInfo.mDescription, formatInfo.mExtensions });
+            if (formatInfo.mFormat == defaultFormatName) {
+               pluginIndex = i;
+               formatIndex = j;
+            }
+         }
+         ++i;
+      }
+   }
+   
+   wxFileName filename;
+   filename.SetPath(defaultDir);
+   
+   if(defaultName.empty())
+      filename.SetName(_("untitled"));
+   else
+      filename.SetFullName(defaultName);
+   
+   if (!filename.HasExt())
+   {
+      auto defext = plugins[pluginIndex]->GetFormatInfo(formatIndex).mExtensions[0].Lower();
+      filename.SetExt(defext);
+   }
+   
+   wxTabTraversalWrapper<FileDialog>::Create(parent,
+                                             title.Translation(),
+                                             defaultDir,
+                                             filename.GetFullName(),
+                                             FileNames::FormatWildcard(fileTypes),
+                                             wxRESIZE_BORDER | wxFD_SAVE | wxFD_OVERWRITE_PROMPT,
+                                             pos, sz, name);
+   
+   Bind(wxEVT_FILECTRL_FILTERCHANGED, &ExportFileDialog::OnFilterChanged, this);
+   SetUserPaneCreator(CreateUserPaneCallback, (wxUIntPtr) this);
+}
+
+int ExportFileDialog::Export(wxWindow* parent,
```

Comment:
`Export` is a misleading name.  This function only chooses a path before export really happens.